### PR TITLE
Sockets: do 0 byte reads to avoid allocating memory for idle connections

### PR DESCRIPTION
**Pull request by tmds**
Implements https://github.com/aspnet/KestrelHttpServer/issues/2334

CC @davidfowl @Tratcher

### CODE REVIEW

File: src/Kestrel.Transport.Sockets/Internal/SocketAwaitable.cs [hunk 1]
```diff
@@ -34,12 +34,18 @@ public int GetResult()
 
             _callback = null;
```
Comment (davidfowl) [inline src/Kestrel.Transport.Sockets/Internal/SocketAwaitable.cs]: We should be assigning _error to a local error variable then setting it to null here.

File: src/Kestrel.Transport.Sockets/Internal/SocketConnection.cs [hunk 2]
```diff
@@ -167,6 +168,11 @@ internal SocketConnection(Socket socket, MemoryPool<byte> memoryPool, PipeSchedu
                     _trace.ConnectionReadFin(ConnectionId);
                     break;
                 }
+                else if (bytesReceived == WouldBlock)
+                {
+                    // WouldBlock
```
Comment (davidfowl) [inline src/Kestrel.Transport.Sockets/Internal/SocketConnection.cs]: This might be worth logging at the trace level.

File: src/Kestrel.Transport.Sockets/Internal/SocketReceiver.cs [hunk 3]
```diff
@@ -12,30 +12,49 @@ public class SocketReceiver
...
+                try
+                {
+#if NETCOREAPP2_1
+                    bytesTransferred = _socket.Receive(_buffer.Span, SocketFlags.None, out errorCode);
+#else
+                    var segment = _buffer.GetArray();
+                    bytesTransferred = _socket.Receive(segment.Array, segment.Offset, segment.Count, SocketFlags.None, out errorCode);
+#endif
+                }
+                catch (ObjectDisposedException)
```
Comment (davidfowl) [inline src/Kestrel.Transport.Sockets/Internal/SocketReceiver.cs]: Is this the only exception that could occur?

**Comment by davidfowl** [comment]
@tmds It also seems like we're seeing a different kind of exception on windows. This kinda applies to the comment above https://github.com/aspnet/KestrelHttpServer/pull/2393#discussion_r174386609

```
xUnit.net 00:00:24.5106913]       Expected: typeof(System.Threading.Tasks.TaskCanceledException)
349[xUnit.net 00:00:24.5108503]       Actual:   typeof(System.IO.IOException): A request to send or receive data was disallowed because the socket had already been shut down in that direction with a previous shutdown call
```

File: src/Kestrel.Transport.Sockets/Internal/SocketConnection.cs (hunk 2 above)
Comment (tmds) [reply src/Kestrel.Transport.Sockets/Internal/SocketConnection.cs]: This should be very rare. The OS says: data is available, and then when we try to receive it says: try again.

File: src/Kestrel.Transport.Sockets/Internal/SocketReceiver.cs (hunk 3 above)
Comment (tmds) [reply src/Kestrel.Transport.Sockets/Internal/SocketReceiver.cs]: Exceptions we get here are due to calling Dispose in SocketConnection. Errors for the receive operation are reported in errorCode.

**Comment by tmds** [comment]
> We need to not allocate any buffer until there is data. We're still calling GetMemory and handing storing it on the SocketReceiver.

Ah yes. GetMemory should be called in OnReadable.

> Ideally the receiver would only return when there is data. We should keep WouldBlock internal to the SocketReceiver so the connection only yields once there's something to do.

I'll look at moving this.

> It also seems like we're seeing a different kind of exception on windows.

There is some exception changing logic in SocketConnection. I think we are hitting a case that should cause ConnectionAbortedException completion, but we see IOException instead.

**Comment by davidfowl** [comment]
> There is some exception changing logic in SocketConnection. I think we are hitting a case that should cause ConnectionAbortedException completion, but we see IOException instead.

I think you want to handle this by catching any exception on read (ExceptionDispatchInfo.Catch()) and passing it to the SocketAwitable.Complete so that GetResult can rethrow it so that SocketConnection can properly handle it.

File: src/Kestrel.Transport.Sockets/Internal/SocketReceiver.cs [hunk 4]
```diff
@@ -12,30 +12,49 @@ public class SocketReceiver
...
         public SocketAwaitable ReceiveAsync(Memory<byte> buffer)
         {
-#if NETCOREAPP2_1
-            _eventArgs.SetBuffer(buffer);
-#else
-            var segment = buffer.GetArray();
-
-            _eventArgs.SetBuffer(segment.Array, segment.Offset, segment.Count);
-#endif
+            _buffer = buffer;
```
Comment (halter73) [inline src/Kestrel.Transport.Sockets/Internal/SocketReceiver.cs]: Why is this a field instead of a local/param?

File: src/Kestrel.Transport.Sockets/Internal/SocketReceiver.cs [hunk 5]
```diff
@@ -12,30 +12,49 @@ public class SocketReceiver
...
             if (!_socket.ReceiveAsync(_eventArgs))
             {
-                _awaitable.Complete(_eventArgs.BytesTransferred, _eventArgs.SocketError);
+                OnReadable();
             }
-
             return _awaitable;
         }
+
+        private void OnReadable()
```
Comment (halter73) [inline src/Kestrel.Transport.Sockets/Internal/SocketReceiver.cs]: Instead of creating this OnReadable callback which is kind of hard to follow being callback base, wouldn't it be cleaner to do something along the lines of the following in SocketConnection.ProcessReceives?

```c#
                await _receiver.WaitForDataAsync();

                // Ensure we have some reasonable amount of buffer space
                var buffer = Input.GetMemory(MinAllocBufferSize);

                var bytesReceived = await _receiver.ReceiveAsync(buffer);
```

This has the additional benefit of delaying the allocation of the buffer until data is available which is a feature I'm pretty sure we want 😉

File: src/Kestrel.Transport.Sockets/Internal/SocketReceiver.cs (hunk 3 above)
Comment (halter73) [reply src/Kestrel.Transport.Sockets/Internal/SocketReceiver.cs]: If we change the code in the way I suggest in my last comment, the ODE will be caught and converted into ConnectionAbortedException as it should without any new code.

File: src/Kestrel.Transport.Sockets/Internal/SocketAwaitable.cs (hunk 1 above)
Comment (halter73) [reply src/Kestrel.Transport.Sockets/Internal/SocketAwaitable.cs]: Good catch. Didn't matter before because all the errors were terminal.

File: src/Kestrel.Transport.Sockets/Internal/SocketConnection.cs (hunk 2 above)
Comment (halter73) [reply src/Kestrel.Transport.Sockets/Internal/SocketConnection.cs]: I'm glad we have someone who has experience with this stuff working on this. It looked to me like we were just rapidly spinning while waiting for more data 🤣

File: src/Kestrel.Transport.Sockets/Internal/SocketReceiver.cs (hunk 5 above)
Comment (tmds) [reply src/Kestrel.Transport.Sockets/Internal/SocketReceiver.cs]: Great suggestion! That will be a lot cleaner!

File: src/Kestrel.Transport.Sockets/Internal/SocketAwaitable.cs (hunk 1 above)
Comment (halter73) [reply src/Kestrel.Transport.Sockets/Internal/SocketAwaitable.cs]: I see that it gets set every time on complete. nvm then.

**Comment by halter73** [comment]
Looks nice. This PR as is seems to have a negligible impact on the pipelined plaintext benchmark but causes about a 4% regression in the JSON benchmark.

#### JSON Benchmark (dev)

```
commit 9901f0f3e4f9bf7053fa7f4b072cb72d67bedb43 (HEAD -> dev, origin/dev, origin/HEAD)
> dotnet build -c Release G:\dev\aspnet\KestrelHttpServer\src\Kestrel.Transport.Sockets; dotnet .\bin\Release\netcoreapp2.0\BenchmarksDriver.dll -n json -j ../Benchmarks/benchmarks.json.json -s http://{windows server} -c http://{linux load client} --webHost KestrelSockets --outputFile G:\dev\aspnet\KestrelHttpServer\src\Kestrel.Transport.Sockets\bin\Release\netcoreapp2.1\Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.dll

[03:58:14.125] RequestsPerSecond:           363136
[03:59:46.337] RequestsPerSecond:           358460
[04:01:12.308] RequestsPerSecond:           366681
```

#### JSON Benchmark (tmds:sockets_zerobyte_receive)

```
commit 2283605b6b286ad105737547df0e6a769a90ae2d (HEAD -> tmds/sockets-zerobyte-receive)
> dotnet build -c Release G:\dev\aspnet\KestrelHttpServer\src\Kestrel.Transport.Sockets; dotnet .\bin\Release\netcoreapp2.0\BenchmarksDriver.dll -n json -j ../Benchmarks/benchmarks.json.json -s http://{windows server} -c http://{linux load client} --webHost KestrelSockets --outputFile G:\dev\aspnet\KestrelHttpServer\src\Kestrel.Transport.Sockets\bin\Release\netcoreapp2.1\Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.dll

[03:34:00.702] RequestsPerSecond:           344616
[03:37:34.402] RequestsPerSecond:           341321
[03:38:37.247] RequestsPerSecond:           345635
```

#### Pipelined Plaintext Benchmark (dev)

```
commit 9901f0f3e4f9bf7053fa7f4b072cb72d67bedb43 (HEAD -> dev, origin/dev, origin/HEAD)
> dotnet build -c Release G:\dev\aspnet\KestrelHttpServer\src\Kestrel.Transport.Sockets; dotnet .\bin\Release\netcoreapp2.0\BenchmarksDriver.dll -n plaintext -j ../Benchmarks/benchmarks.plaintext.json -s http://{windows server} -c http://{linux load client} --webHost KestrelSockets --outputFile G:\dev\aspnet\KestrelHttpServer\src\Kestrel.Transport.Sockets\bin\Release\netcoreapp2.1\Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.dll

[04:03:22.783] RequestsPerSecond:           1772742
[04:04:53.505] RequestsPerSecond:           1791719
[04:06:18.725] RequestsPerSecond:           1832628
```

#### Pipelined Plaintext Benchmark (mds:sockets_zerobyte_receive)

```
commit 2283605b6b286ad105737547df0e6a769a90ae2d (HEAD -> tmds/sockets-zerobyte-receive)
> dotnet build -c Release G:\dev\aspnet\KestrelHttpServer\src\Kestrel.Transport.Sockets; dotnet .\bin\Release\netcoreapp2.0\BenchmarksDriver.dll -n plaintext -j ../Benchmarks/benchmarks.plaintext.json -s http://{windows server} -c http://{linux load client} --webHost KestrelSockets --outputFile G:\dev\aspnet\KestrelHttpServer\src\Kestrel.Transport.Sockets\bin\Release\netcoreapp2.1\Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.dll

[04:07:45.877] RequestsPerSecond:           1784669
[04:09:12.727] RequestsPerSecond:           1785730
[04:10:39.937] RequestsPerSecond:           1770847
```

**Comment by davidfowl** [comment]
There's going to be some overhead changing the buffers from empty to a real buffer every time. I'm not sure how big of a deal that is though. Also though the code is simpler it might not be optimal to do 2 of these async reads every time. Also it might be possible to completely avoid waiting if Socket.Available > 0 (we'd need to measure). I also want to do some memory tests.

/cc @BrennanConroy this should go nicely with your 0 byte websocket receive tests.

**Comment by tmds** [comment]
> Also though the code is simpler it might not be optimal to do 2 of these async reads every time.

When doing an async socket operation, corefx first tries to do the operation synchronously.

So when there is data available, the calls looks like:
```
WaitForDataAsync
- check available (recvmsg 1 byte peek)
ReceiveAsync
- read (recvmsg)
```
If there is no data it looks like:
```
WaitForDataAsync
- check available
- wait for data (epoll)
- check available (*)
ReceiveAsync
- read
```
(*) this check is due to the corefx implementation to handle non zero byte reads. For zero byte reads, it should be possible to eliminate this.

> Also it might be possible to completely avoid waiting if Socket.Available > 0

Socket.Available adds an additional system call (ioctl FIONREAD), this will perform worse.

**Comment by tmds** [comment]
> For zero byte reads, it should be possible to eliminate this.

https://github.com/dotnet/corefx/pull/28098

I wonder if it will cut some of the overhead on the JSON benchmark.

**Comment by tmds** [comment]
> dotnet/corefx#28098

This has been merged. Once the change ripples through, we can do an new benchmark.

**Comment by davidfowl** [comment]
@tmds We can run it before it ripples through.

**Comment by tmds** [comment]
@halter73 can you please benchmark again and see if eliminating the syscall reduced the json perf regression?

**Comment by davidfowl** [comment]
@dotnet-bot test OSX 10.12 Release Build

**Comment by halter73** [comment]
@tmds Do you know the first runtime build number that includes your dotnet/corefx#28098 changes?

**Comment by halter73** [comment]
I just ran the JSON tests on linux with version 2.1.0-preview3-26322-05 of the runtime which was built using [this corefx commit](https://github.com/dotnet/corefx/commits/969541fb79ccacbe07965e08e3b32a9d5ecb19bb). Numbers look much better than they do on Windows. There appears to be about a 1% RPS regression on Linux instead of the 4% RPS regression we've seen on Windows. Do you think we can make a similar optimization on Windows?

### Linux JSON Baseline

```
G:\dev\aspnet\benchmarks\src\BenchmarksDriver [dev ≡]> dotnet build -c Release G:\dev\aspnet\KestrelHttpServer\src\Kestrel.Transport.Sockets; dotnet .\bin\Release\netcoreapp2.0\BenchmarksDriver.dll -n json -j ../Benchmarks/benchmarks.json.json -s http://{linux server}:5001 -c http://{linux load client}:5002 --webHost KestrelSockets --runtimeVersion edge --outputFile G:\dev\aspnet\KestrelHttpServer\src\Kestrel.Transport.Sockets\bin\Release\netcoreapp2.1\Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.dll
[05:47:22.815] RequestsPerSecond:           315455
[05:47:22.815] Latency on load (ms):        1.2
[05:47:22.816] Max CPU (%):                 95
[05:47:22.816] WorkingSet (MB):             412
[05:47:22.816] Startup Main (ms):           424
[05:47:22.816] First Request (ms):          309.6
[05:47:22.817] Latency (ms):                0.6
[05:47:22.817] Total Requests:              4,763,305
[05:47:22.817] Duration: (ms)               15100
[05:47:22.817] Socket Errors:               0
[05:47:22.817] Bad Responses:               0
[05:47:22.817] SDK:                         2.2.0-preview1-007522
[05:47:22.817] Runtime:                     2.1.0-preview3-26322-05
[05:47:22.817] ASP.NET Core:                2.1.0-preview3-32010

G:\dev\aspnet\benchmarks\src\BenchmarksDriver [dev ≡]> dotnet build -c Release G:\dev\aspnet\KestrelHttpServer\src\Kestrel.Transport.Sockets; dotnet .\bin\Release\netcoreapp2.0\BenchmarksDriver.dll -n json -j ../Benchmarks/benchmarks.json.json -s http://{linux server}:5001 -c http://{linux load client}:5002 --webHost KestrelSockets --runtimeVersion edge --outputFile G:\dev\aspnet\KestrelHttpServer\src\Kestrel.Transport.Sockets\bin\Release\netcoreapp2.1\Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.dll
[05:48:19.184] RequestsPerSecond:           313727
[05:48:19.184] Latency on load (ms):        1
[05:48:19.185] Max CPU (%):                 95
[05:48:19.185] WorkingSet (MB):             409
[05:48:19.185] Startup Main (ms):           419
[05:48:19.185] First Request (ms):          308.5
[05:48:19.185] Latency (ms):                0.8
[05:48:19.186] Total Requests:              4,737,110
[05:48:19.186] Duration: (ms)               15100
[05:48:19.186] Socket Errors:               0
[05:48:19.186] Bad Responses:               0
[05:48:19.186] SDK:                         2.2.0-preview1-007522
[05:48:19.186] Runtime:                     2.1.0-preview3-26322-05
[05:48:19.186] ASP.NET Core:                2.1.0-preview3-32010

G:\dev\aspnet\benchmarks\src\BenchmarksDriver [dev ≡]> dotnet build -c Release G:\dev\aspnet\KestrelHttpServer\src\Kestrel.Transport.Sockets; dotnet .\bin\Release\netcoreapp2.0\BenchmarksDriver.dll -n json -j ../Benchmarks/benchmarks.json.json -s http://{linux server}:5001 -c http://{linux load client}:5002 --webHost KestrelSockets --runtimeVersion edge --outputFile G:\dev\aspnet\KestrelHttpServer\src\Kestrel.Transport.Sockets\bin\Release\netcoreapp2.1\Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.dll
[05:49:10.735] RequestsPerSecond:           314169
[05:49:10.735] Latency on load (ms):        1.1
[05:49:10.735] Max CPU (%):                 97
[05:49:10.736] WorkingSet (MB):             409
[05:49:10.736] Startup Main (ms):           391
[05:49:10.736] First Request (ms):          307.8
[05:49:10.736] Latency (ms):                0.8
[05:49:10.736] Total Requests:              4,743,821
[05:49:10.736] Duration: (ms)               15100
[05:49:10.736] Socket Errors:               0
[05:49:10.736] Bad Responses:               0
[05:49:10.737] SDK:                         2.2.0-preview1-007522
[05:49:10.737] Runtime:                     2.1.0-preview3-26322-05
[05:49:10.737] ASP.NET Core:                2.1.0-preview3-32010
[05:49:10.737] Stopping scenario Json on benchmark server...
[05:49:12.747] Deleting scenario Json on benchmark server...
```

### Linux JSON w/ this PR

```
G:\dev\aspnet\benchmarks\src\BenchmarksDriver [dev ≡]> dotnet build -c Release G:\dev\aspnet\KestrelHttpServer\src\Kestrel.Transport.Sockets; dotnet .\bin\Release\netcoreapp2.0\BenchmarksDriver.dll -n json -j ../Benchmarks/benchmarks.json.json -s http://{linux server}:5001 -c http://{linux load client}:5002 --webHost KestrelSockets --runtimeVersion edge --outputFile G:\dev\aspnet\KestrelHttpServer\src\Kestrel.Transport.Sockets\bin\Release\netcoreapp2.1\Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.dll
[05:51:38.100] RequestsPerSecond:           311179
[05:51:38.101] Latency on load (ms):        1.2
[05:51:38.101] Max CPU (%):                 95
[05:51:38.101] WorkingSet (MB):             408
[05:51:38.102] Startup Main (ms):           388
[05:51:38.102] First Request (ms):          311.7
[05:51:38.102] Latency (ms):                0.7
[05:51:38.102] Total Requests:              4,698,551
[05:51:38.102] Duration: (ms)               15100
[05:51:38.102] Socket Errors:               0
[05:51:38.102] Bad Responses:               0
[05:51:38.103] SDK:                         2.2.0-preview1-007522
[05:51:38.103] Runtime:                     2.1.0-preview3-26322-05
[05:51:38.103] ASP.NET Core:                2.1.0-preview3-32010

G:\dev\aspnet\benchmarks\src\BenchmarksDriver [dev ≡]> dotnet build -c Release G:\dev\aspnet\KestrelHttpServer\src\Kestrel.Transport.Sockets; dotnet .\bin\Release\netcoreapp2.0\BenchmarksDriver.dll -n json -j ../Benchmarks/benchmarks.json.json -s http://{linux server}:5001 -c http://{linux load client}:5002 --webHost KestrelSockets --runtimeVersion edge --outputFile G:\dev\aspnet\KestrelHttpServer\src\Kestrel.Transport.Sockets\bin\Release\netcoreapp2.1\Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.dll
[05:52:34.953] RequestsPerSecond:           313176
[05:52:34.953] Latency on load (ms):        1.2
[05:52:34.953] Max CPU (%):                 96
[05:52:34.953] WorkingSet (MB):             408
[05:52:34.954] Startup Main (ms):           422
[05:52:34.954] First Request (ms):          304.9
[05:52:34.954] Latency (ms):                0.9
[05:52:34.954] Total Requests:              4,728,732
[05:52:34.954] Duration: (ms)               15100
[05:52:34.954] Socket Errors:               0
[05:52:34.954] Bad Responses:               0
[05:52:34.954] SDK:                         2.2.0-preview1-007522
[05:52:34.955] Runtime:                     2.1.0-preview3-26322-05
[05:52:34.955] ASP.NET Core:                2.1.0-preview3-32010

G:\dev\aspnet\benchmarks\src\BenchmarksDriver [dev ≡]> dotnet build -c Release G:\dev\aspnet\KestrelHttpServer\src\Kestrel.Transport.Sockets; dotnet .\bin\Release\netcoreapp2.0\BenchmarksDriver.dll -n json -j ../Benchmarks/benchmarks.json.json -s http://{linux server}:5001 -c http://{linux load client}:5002 --webHost KestrelSockets --runtimeVersion edge --outputFile G:\dev\aspnet\KestrelHttpServer\src\Kestrel.Transport.Sockets\bin\Release\netcoreapp2.1\Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.dll
[05:53:27.612] RequestsPerSecond:           307501
[05:53:27.613] Latency on load (ms):        1.2
[05:53:27.613] Max CPU (%):                 96
[05:53:27.613] WorkingSet (MB):             413
[05:53:27.614] Startup Main (ms):           409
[05:53:27.614] First Request (ms):          304
[05:53:27.614] Latency (ms):                0.9
[05:53:27.614] Total Requests:              4,643,170
[05:53:27.614] Duration: (ms)               15100
[05:53:27.614] Socket Errors:               0
[05:53:27.614] Bad Responses:               0
[05:53:27.615] SDK:                         2.2.0-preview1-007522
[05:53:27.615] Runtime:                     2.1.0-preview3-26322-05
[05:53:27.615] ASP.NET Core:                2.1.0-preview3-32010
```

**Comment by tmds** [comment]
> I just ran the JSON tests on linux with version 2.1.0-preview3-26322-05 of the runtime which was built using this corefx commit. Numbers look much better than they do on Windows. There appears to be about a 1% RPS regression on Linux instead of the 4% RPS regression we've seen on Windows. Do you think we can make a similar optimization on Windows?

I had a look in corefx Windows socket implementation and I didn't see a superfluous system call like there was for Unix.

**Comment by davidfowl** [comment]
@tmds can you rebase on the latest dev? We have new corefx bits and we'd like to re-measure. If you want I can take it over from you (if you don't have time).

**Comment by tmds** [comment]
@davidfowl, I merged with dev.

**Comment by davidfowl** [comment]
@dotnet-bot test this please

**Comment by tmds** [comment]
@dotnet-bot test this please

**Comment by davidfowl** [comment]
@tmds closing in favor of https://github.com/aspnet/KestrelHttpServer/pull/2502

**Comment by tmds** [comment]
ok. fwiw I could have squashed these commits for you.

**Comment by davidfowl** [comment]
@tmds my bad. I left this PR open for too long and we should really have merged it a while ago...
